### PR TITLE
fix(is_external_link): handle invalid url

### PR DESCRIPTION
**Pull request by SukkaW**
Details can be found at https://github.com/hexojs/hexo/issues/4130

This PR introduce a try block and treat invalid URL as internal link (return `false`).

### CODE REVIEW

**Comment by SukkaW** [comment]
According to [travis ci build](https://travis-ci.com/sukkaw/hexo/builds/149168430), no performance impact detected.

**Comment by SukkaW** [comment]
`is-absolute-url` is also added for better performance.

Since it is a widely used packages on npm so I believe it is reliable enough.

File: lib/is_external_link.js [hunk 1]
```diff
@@ -13,11 +14,19 @@ const cache = new Cache();
 
 function isExternalLink(input, sitehost, exclude) {
   return cache.apply(`${input}-${sitehost}-${exclude}`, () => {
-    sitehost = parse(sitehost).hostname || sitehost;
+    if (!isAbsoluteUrl(input)) return false;
```
Comment (curbengh) [inline lib/is_external_link.js]: `isAbsoluteUrl` matches [any](https://github.com/sindresorhus/is-absolute-url/blob/8a57ccec064d925640860674cd32e7182eeb3596/index.js#L15) protocol. For our use case, `/^http(s)?:\/\//.test(input)` alone should be sufficient.

Comment (SukkaW) [reply lib/is_external_link.js]: @curbengh Agreed. I will update to it then.

Comment (SukkaW) [reply lib/is_external_link.js]: Updated to handle `http://`, `https://` and **`//`** prefixed url.

File: lib/is_external_link.js [hunk 2]
```diff
@@ -13,11 +13,21 @@ const cache = new Cache();
...
 
     if (!sitehost) return false;
-    // handle relative url
-    const data = new URL(input, `http://${sitehost}`);
+
+    // handle relative url and invalid url
+    let data;
+    try {
+      data = new URL(input, `http://${sitehost}`);
+    } catch (e) { }
```
Comment (jiangtj) [inline lib/is_external_link.js]: How about `hexo.log.warn(...)` to alert the user that there has an invalid URL?

Comment (SukkaW) [reply lib/is_external_link.js]: Since hexo-util has no `hexo-log`'s dependency, there is no way we can use `hexo.log.warn` in hexo-util.

**Pull request merged**